chain/tsresolver/tipset_resolver.go

```diff
@@ -0,0 +1,164 @@
+
+	case EthBlockSelectorPending:
+		return tsr.loader.GetHeaviestTipSet(), nil
+
+	case EthBlockSelectorLatest:
+		// head - 1 because we're always one behind for Eth compatibility due to deferred execution
+		return tsr.loader.LoadTipSet(ctx, tsr.loader.GetHeaviestTipSet().Parents())
+
+	case EthBlockSelectorSafe, EthBlockSelectorFinalized:
+		defaultDelay := policy.ChainFinality
+		if selector == EthBlockSelectorSafe {
+			defaultDelay = ethtypes.SafeEpochDelay
+		}
+
+		head := tsr.loader.GetHeaviestTipSet()
+		latestHeight := head.Height() - 1 // always one behind for Eth compatibility due to deferred execution
+		defaultHeight := latestHeight - defaultDelay
+
+		if f3TipSet, err := tsr.getF3FinalizedTipSet(ctx); err != nil {
+			return nil, err
+		} else if f3TipSet != nil && f3TipSet.Height() > defaultHeight {
+			// return the parent of the finalized tipset (deferred execution, eth cares about t)
+			return tsr.loader.LoadTipSet(ctx, f3TipSet.Parents())
+		} // else F3 is disabled, not running, or behind the default safe or finalized height
+
+		ts, err := tsr.loader.GetTipsetByHeight(ctx, defaultHeight, head, true)
+		if err != nil {
+			return nil, xerrors.Errorf("loading tipset at height %v: %w", defaultHeight, err)
+		}
+		return ts, nil
+
+	default:
+		// likely an 0x hex block number or a decimal block number
+		return tsr.resolveEthBlockNumberSelector(ctx, selector, strict)
+	}
+}
+
+func (tsr *tipSetResolver) getF3FinalizedTipSet(ctx context.Context) (*types.TipSet, error) {
+	if !tsr.f3.IsEnabled() {
+		return nil, nil
+	}
+	if running, _ := tsr.f3.IsRunning(); !running {
+		return nil, nil
+	}
+
+	cert, err := tsr.f3.GetLatestCert(ctx)
+	if err != nil {
+		log.Debugf("loading latest F3 certificate: %s", err)
+		return nil, nil
+	}
+
+	tsk, err := types.TipSetKeyFromBytes(cert.ECChain.Head().Key)
+	if err != nil {
+		return nil, xerrors.Errorf("decoding tipset key reported by F3: %w", err)
+	}
+
+	finalizedTipSet, err := tsr.loader.LoadTipSet(ctx, tsk)
+	if err != nil {
+		return nil, xerrors.Errorf("loading tipset reported as finalized by F3 %s: %w", tsk, err)
+	}
+
+	return finalizedTipSet, nil
+}
+
+func (tsr *tipSetResolver) resolveEthBlockNumberSelector(ctx context.Context, selector string, strict bool) (*types.TipSet, error) {
+	var num ethtypes.EthUint64
+	if err := num.UnmarshalJSON([]byte(`"` + selector + `"`)); err != nil {
+		return nil, xerrors.Errorf("cannot parse block number: %v", err)
+	}
+
+	head := tsr.loader.GetHeaviestTipSet()
+	if abi.ChainEpoch(num) > head.Height()-1 {
+		return nil, errors.New("requested a future epoch (beyond 'latest')")
+	}
+
+	ts, err := tsr.loader.GetTipsetByHeight(ctx, abi.ChainEpoch(num), head, true)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get tipset at height: %v", num)
+	}
+
+	if strict && ts.Height() != abi.ChainEpoch(num) {
+		return nil, api.NewErrNullRound(abi.ChainEpoch(num))
+	}
```

Comment:
Some of our APIs can't be sloppy and if you request a specific block then you should get it; we've standardised returning an `api.ErrNullRound` that passes through the RPC layer in this case. It's a departure from expected Ethereum tooling behaviour simply because they don't have null rounds to deal with, but we've decided that it's better to return an error than an actual `null` or an alternative tipset, and consumers have adapted to this. There's an itest that asserts this behaviour holds: https://github.com/filecoin-project/lotus/blob/dbf41383b37f1155dd0a28b10da2909226adb171/itests/fevm_test.go#L1664